gobblin-yarn/src/main/java/gobblin/yarn/GobblinApplicationMaster.java

```diff
@@ -0,0 +1,453 @@
+      this.metricRegistry.register(MetricRegistry.name(prefix, entry.getKey()), entry.getValue());
+    }
+  }
+
+  private void addShutdownHook() {
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+
+      @Override
+      public void run() {
+        GobblinApplicationMaster.this.stop();
+      }
+    });
+  }
+
+  private void sendShutdownRequest() {
+    Criteria criteria = new Criteria();
+    criteria.setInstanceName("%");
+    criteria.setResource("%");
+    criteria.setPartition("%");
+    criteria.setPartitionState("%");
+    criteria.setRecipientInstanceType(InstanceType.PARTICIPANT);
+    criteria.setDataSource(Criteria.DataSource.LIVEINSTANCES);
+    criteria.setSessionSpecific(true);
+
+    Message shutdownRequest = new Message(Message.MessageType.SHUTDOWN,
+        HelixMessageSubTypes.WORK_UNIT_RUNNER_SHUTDOWN.toString().toLowerCase() + UUID.randomUUID().toString());
+    shutdownRequest.setMsgSubType(HelixMessageSubTypes.WORK_UNIT_RUNNER_SHUTDOWN.toString());
+    shutdownRequest.setMsgState(Message.MessageState.NEW);
+
+    int messagesSent = this.helixManager.getMessagingService().send(criteria, shutdownRequest);
+    if (messagesSent == 0) {
+      LOGGER.error(String.format("Failed to send the %s message to the participants", shutdownRequest.getMsgSubType()));
+    }
+  }
+
+  /**
+   * A custom implementation of {@link LiveInstanceChangeListener}.
+   */
+  private static class GobblinLiveInstanceChangeListener implements LiveInstanceChangeListener {
+
+    @Override
+    public void onLiveInstanceChange(List<LiveInstance> liveInstances, NotificationContext changeContext) {
+      for (LiveInstance liveInstance : liveInstances) {
+        LOGGER.info("Live Helix participant instance: " + liveInstance.getInstanceName());
+      }
+    }
+  }
+
+  /**
+   * A custom {@link MessageHandlerFactory} for {@link MessageHandler}s that handle messages of type
+   * {@link org.apache.helix.model.Message.MessageType#SHUTDOWN} for shutting down the controller.
+   */
+  private class ControllerShutdownMessageHandlerFactory implements MessageHandlerFactory {
+
+    @Override
+    public MessageHandler createHandler(Message message, NotificationContext context) {
+      return new ControllerShutdownMessageHandler(message, context);
+    }
+
+    @Override
+    public String getMessageType() {
+      return Message.MessageType.SHUTDOWN.toString();
+    }
+
+    @Override
+    public void reset() {
+
+    }
+
+    /**
+     * A custom {@link MessageHandler} for handling messages of sub type
+     * {@link HelixMessageSubTypes#APPLICATION_MASTER_SHUTDOWN}.
+     */
+    private class ControllerShutdownMessageHandler extends MessageHandler {
+
+      public ControllerShutdownMessageHandler(Message message, NotificationContext context) {
+        super(message, context);
+      }
+
+      @Override
+      public HelixTaskResult handleMessage() throws InterruptedException {
+        String messageSubType = this._message.getMsgSubType();
```

Comment:
`_message` is defined in `MessageHandler` as part of Helix so I'm just using it. it's impossible to use `this.` in `run` below.